File: librtEdge/cpp/SysAgent.cpp
/******************************************************************************
*
*  SysAgent.cpp
*     Connectionless publisher agent to monitor CPU and Disk stats
*
*  REVISION HISTORY:
*     13 JUL 2017 jcs  Created.
*
*  (c) 1994-2017 Gatea Ltd.
******************************************************************************/
#include <librtEdge.h>

// Hard-coded

static int _nameFid =    3;
static int _cpuFid  =    6;
static int _seqFid  = 1021; // SEQNUM

using namespace RTEDGE;

// Locals

const char *_pSigLog = NULL;


/////////////////////////////////
// Support Functions
/////////////////////////////////
#if !defined(WIN32)
#include <assert.h>
#include <errno.h>
#include <signal.h>

void forkAndSig( bool bFork, void (*handler)( int ), int *sigs )
{
   int              i, pid;
   sigset_t         sigSet;
   struct sigaction s, lwc;

   // fork() if !StdOut

   if ( bFork ) {
      if ( (pid=::fork()) < 0 ) { // Error
         ::fprintf( stdout, "\nfork() ERROR:  %s\n", ::strerror( errno ) );
         exit( 1 );
      }
      else if ( pid > 0 )   // Parent
         exit( 0 );

      // Child:
      // 1) Ignore terminal stop signals (BSD).
      // 2) Process is session leader
      // 3) Disassociate from process group
      // 4) Change current dir to file system root so
      //    current filesystem may be unmounted.
      // 5) Clear file mode creation mask
      // 6) Ignore SIGCLD, allowing dead children to bypass
      //    zombie state and free up process table slot.

      int ttSigs[] = { SIGTTOU, SIGTTIN, SIGTSTP, 0 };

      ::sigemptyset( &sigSet );
      for ( i=0; ttSigs[i]; ::sigaddset( &sigSet, ttSigs[i++] ) );
      ::sigprocmask( SIG_UNBLOCK, &sigSet, (sigset_t *)0 );
      ::setpgrp();
/*
      ::chdir( "/" );
      ::umask( 0 );
 */
      ::sigemptyset( &sigSet );
      ::sigaddset( &sigSet, SIGHUP );
      ::sigaddset( &sigSet, SIGCLD );
      ::sigprocmask( SIG_UNBLOCK, &sigSet, (sigset_t *)0 );

      // Really get rid of controlling terminal

      if ( (pid=::fork()) < 0 ) { // Error
         ::fprintf( stdout, "\nfork() ERROR:  %s\n", ::strerror( errno ) );
         exit( 1 );
      }
      else if ( pid > 0 ) {       // Parent
         ::fprintf( stdout, "\nChild PID = %d\n", pid );
         exit( 0 );
      }
   }

   // Signal handlers

   for ( i=0; sigs[i]; i++ ) {
      ::sigemptyset( &(s.sa_mask) );
      s.sa_handler = handler;
      s.sa_flags   = 0;
      ::sigaction( sigs[i], &s, &lwc );
   }
}

void sigHandler( int num )
{
   switch( num ) {
      case SIGTERM: _pSigLog = "SIGTERM"; break;
      case SIGINT:  _pSigLog = "SIGINT";  break;
//      case SIGSEGV: _pSigLog = "SIGSEGV"; break;
      case SIGSEGV: assert( 0 ); break;
   }
}

#endif // !defined(WIN32)

class MyChannel : public PubChannel
{
   ////////////////////////////////
   // Members
   ////////////////////////////////
private:
   Mutex      _mtx;
   CPUStats  *_cs;
   DiskStats *_ds;
   int        _RTL;
   char       _host[K];

   ////////////////////////////////
   // Constructor
   ////////////////////////////////
public:
   MyChannel( const char *pPub, bool bCPU ) :
      PubChannel( pPub ),
      _mtx(),
      _cs( (CPUStats *)0 ),
      _ds( (DiskStats *)0 ),
      _RTL( 1 )
   {
      strcpy( _host, "Undefined" );
      if ( bCPU )
         _cs = new CPUStats();
      else
         _ds = new DiskStats();
   }

   ~MyChannel()
   {
      if ( _cs )
         delete _cs;
      if ( _ds )
         delete _ds;
   }


   ////////////////////////////////
   // Operations
   ////////////////////////////////
public:
   char *SetHost()
   {
      /*
       * WIN32 requires WinSock to load
       * Call this AFTER Start() or StartConnectionless()
       */
      ::gethostname( _host, sizeof( _host ) );
      return _host;
   }

   int PublishAll()
   {
      int nr;

      if ( _cs ) nr = _Publish( *_cs );
      else       nr = _Publish( *_ds );
      _RTL++;
      return nr;
   }

   int _Publish( CPUStats &cs )
   {
      Locker      lck( _mtx );
      Update     &u = upd();
      ::OSCpuStat c;
      char        tkr[K];
      bool        bImg;
      int         i, id, n, fid;

      n    = cs.Snap();
      bImg = ( _RTL == 1 );
      for ( i=0; i<n; i++ ) {
         id = i+1;
         sprintf( tkr, "%s.%d", _host, i );
         u.Init( tkr, (void *)id, bImg );
         c   = cs.Get( i );
         fid = _cpuFid;
         u.AddField( fid++, c._us );
         u.AddField( fid++, c._sy );
         u.AddField( fid++, c._ni );
         u.AddField( fid++, c._id );
         u.AddField( fid++, c._wa );
         u.AddField( fid++, c._si );
         u.AddField( fid++, c._st );
         u.AddField( _seqFid, _RTL );
         u.Publish();
      }
      return n;
   }

   int _Publish( DiskStats &ds )
   {
      Locker       lck( _mtx );
      Update      &u = upd();
      ::OSDiskStat d;
      char         tkr[K];
      bool         bImg;
      int          i, id, n, fid;

      n    = ds.Snap();
      bImg = ( _RTL == 1 );
      for ( i=0; i<n; i++ ) {
         id = i+1;
         sprintf( tkr, "%s.%d", _host, i );
         u.Init( tkr, (void *)id, bImg );
         d   = ds.Get( i );
         fid = _cpuFid;
         u.AddField( _nameFid, d._diskName );
         u.AddField( fid++, d._nRdSec );
         u.AddField( fid++, d._nWrSec );
         u.AddField( fid++, d._nIoMs );
         u.AddField( _seqFid, _RTL );
         u.Publish();
      }
      return n;
   }


   ////////////////////////////////
   // Asynchronous Callbacks
   ////////////////////////////////
public:
   virtual void OnConnect( const char *msg, bool bUP )
   {
      const char *pUp = bUP ? "UP" : "DOWN";

      ::fprintf( stdout, "CONN %s : %s\n", pUp, msg );
      ::fflush( stdout );
   }

   virtual void OnSymListQuery( int nSym )
   {
      ::fprintf( stdout, "OnSymListQuery( %d )\n", nSym );
      PublishAll();
   }

   virtual void OnRefreshImage( const char *tkr, void *arg )
   {
      ::fprintf( stdout, "OnRefreshImage( %s )\n", tkr );
      ::fflush( stdout );
#ifdef TODO_REFRESH
      if ( (it=_wl.find( s )) != _wl.end() ) {
         w = (*it).second;
         PubTkr( *w );
      }
#endif // TODO_REFRESH
   }


   ////////////////////////////////////
   // PubChannel Interface
   ////////////////////////////////////
protected:
   virtual Update *CreateUpdate()
   {
      return new Update( *this );
   }

}; // class MyChannel


/////////////////
// Helpers
/////////////////
void ProgressBar( int i )
{
   if ( i && !(i%100) )
      printf( "%d", i/100 );
   else if ( i && !(i%50) )
      printf( "+" );
   else if ( i && !(i%10) )
      printf( "."  );
   fflush( stdout );
}


////////////////////////////////////////////
// Do it baby!!
////////////////////////////////////////////
int main( int argc, char **argv )
{
   const char *pSvr, *pPub, *pc;
   double      tSlp;
   int         i;
   bool        bCPU;
#if !defined(WIN32) 
   static int  sigs[] = { SIGTERM, SIGINT, SIGSEGV, 0 };
#endif // !defined(WIN32) 

   // Quickie check

   if ( argc > 1 && !::strcmp( argv[1], "--version" ) ) {
      printf( "%s\n", rtEdge::Version() );
      return 0;
   }

   // cmd-line args

   if ( argc < 4 ) {
      pc = "Usage: %s <hosts> <Svc> <pubTmr> [<bDisk>]";
      printf( pc, argv[0] );
      printf( "; Exitting ...\n" );
      return 0;
   }
   pSvr = argv[1];
   pPub = argv[2];
   tSlp = atof( argv[3] );
   bCPU = ( argc >= 4 );

   MyChannel   pub( pPub, bCPU );

   ::fprintf( stdout, "%s\n", pub.Version() );
   ::fprintf( stdout, "%s\n", pub.StartConnectionless( pSvr ) );
   ::fprintf( stdout, "Publish every %.1fs for %s\n", tSlp, pub.SetHost() );
   ::fflush( stdout );
#if !defined(WIN32) 
   forkAndSig( false, sigHandler, sigs );
#endif // !defined(WIN32) 
   for ( i=0; !_pSigLog; i++ ) {
      pub.Sleep( tSlp );
      pub.PublishAll();
      ProgressBar( i );
   }

   // Clean up

   ::fprintf( stdout, "Cleaning up ...\n" ); ::fflush( stdout );
   pub.Stop();
   printf( "Done!!\n " );
   return 1;
}

File: librtEdge/CLI/examples/SysAgent/SysAgent.cs
/******************************************************************************
*
*  SysAgent.cs
*     Connectionless publisher agent to monitor CPU and Disk stats
*
*  REVISION HISTORY:
*     13 JUL 2017 jcs  Created.
*
*  (c) 1994-2017 Gatea Ltd.
******************************************************************************/
using System;
using System.IO;
using System.Collections;
using System.Net;
using System.Threading;
using System.Runtime.InteropServices;
using librtEdge;

////////////////////////
// System CPU / Disk Monitor
////////////////////////
class SysAgent : rtEdgePublisher
{
   ////////////////////////////////
   // Class-wide Members
   ////////////////////////////////
   static int _nameFid =    3;
   static int _cpuFid  =    6;
   static int _seqFid  = 1021; // SEQNUM

   //////////////
   // Members
   //////////////
   private CPUStats               _cpu;
   private DiskStats              _disk;
   private String                 _host;
   private int                    _RTL;
   private int                    _iTmr;
   private TimerCallback          _cbk;
   private System.Threading.Timer _tmr;
 
   ////////////////////////////////
   // Constructor
   ////////////////////////////////
   public SysAgent( string pSvr, 
                    string pPub, 
                    double tSnap,
                    bool   bDisk ) :
      base( pSvr, pPub )
   {
      String fmt;

      // 1) CPU / Disk Stats

      _cpu  = bDisk ? null : new CPUStats();
      _disk = bDisk ? new DiskStats() : null;
      _host = Dns.GetHostName();
      _RTL  = 1;

      // 2) Connectionless Publisher

      PubStartConnectionless();

      // 3) Snapshot Timer

      _iTmr = Convert.ToInt32( 1000.0 * tSnap );
      _cbk  = new TimerCallback( Snap );
      _tmr  = new System.Threading.Timer( _cbk, null, _iTmr, _iTmr );
      fmt   = "{0} to udp:{1} every {2}ms for {3}";
      Console.WriteLine( fmt, pPub, pSvr, _iTmr, _host );
   }


   ////////////////////////////////
   // rtEdgePublisher Interface
   ////////////////////////////////
   public override void OnConnect( string pConn, rtEdgeState state )
   {
      bool   bUp = ( state == rtEdgeState.edg_up );
      string pUp = bUp ? "UP" : "DOWN";

      Console.WriteLine( "PUB-CONN {0} : {1}", pConn, pUp );
   }

   public override void OnOpen( string tkr, IntPtr arg )
   {
      Console.WriteLine( "OPEN {0}", tkr );
   }

   public override void OnClose( string tkr )
   {
      Console.WriteLine( "CLOSE {0}", tkr );
   }

   public override void OnSymListQuery( int nSym )
   {
      Console.WriteLine( "OnSymListQuery( {0} )", nSym );
   }

   public override void OnRefreshImage( String tkr, IntPtr StreamID )
   {
      Console.WriteLine( "OnRefreshImage( {0} )", tkr );
   }


   #region Timer Handler
   //////////////////////////////
   // Timer Handler
   //////////////////////////////
   private void Snap( object data )
   {
      if ( _cpu != null )
         _Publish( _cpu );
      if ( _disk != null )
         _Publish( _disk );
      _RTL += 1;
   }

   private void _Publish( CPUStats cs )
   {
      rtEdgePubUpdate u;
      String          tkr;
      OSCpuStat       c;
      bool            bImg;
      int             i, id, n, fid;

      n    = cs.Snap();
      bImg = ( _RTL == 1 );
      for ( i=0; i<n; i++ ) {
         id  = i+1;
         tkr = _host + "." + id.ToString();
         u   = new rtEdgePubUpdate( this, tkr, (IntPtr)i, bImg );
         c   = cs.Get( i );
         fid = _cpuFid;
         u.AddFieldAsDouble( fid++, c._us );
         u.AddFieldAsDouble( fid++, c._sy );
         u.AddFieldAsDouble( fid++, c._ni );
         u.AddFieldAsDouble( fid++, c._id );
         u.AddFieldAsDouble( fid++, c._wa );
         u.AddFieldAsDouble( fid++, c._si );
         u.AddFieldAsDouble( fid++, c._st );
         u.AddFieldAsInt32( _seqFid, _RTL );
         u.Publish();
      }
   }

   private void _Publish( DiskStats ds )
   {
      rtEdgePubUpdate u;
      String          tkr;
      OSDiskStat      d;
      bool            bImg;
      int             i, id, n, fid;

      n    = ds.Snap();
      bImg = ( _RTL == 1 );
      for ( i=0; i<n; i++ ) {
         id  = i+1;
         tkr = _host + "." + id.ToString();
         u   = new rtEdgePubUpdate( this, tkr, (IntPtr)i, bImg );
         d   = ds.Get( i );
         fid = _cpuFid;
         u.AddFieldAsString( _nameFid, d.Name() );
         u.AddFieldAsDouble( fid++, d._nRdSec );
         u.AddFieldAsDouble( fid++, d._nWrSec );
         u.AddFieldAsDouble( fid++, d._nIoMs );
         u.AddFieldAsInt32( _seqFid, _RTL );
         u.Publish();
      }
   }

   #endregion
 

   //////////////////////////////
   // main()
   //////////////////////////////
   static int Main( string[] args )
   {
      try {
         SysAgent app;
         String   pSvr, pPub;
         double   tTmr;
         int      argc;
         bool     bDsk;

         // Quickie Check

         argc = args.Length;
         if ( argc > 0 && ( args[0] == "--version" ) ) {
            Console.WriteLine( rtEdge.Version() );
            return 0;
         }
         if ( ( argc > 0 && ( args[0] == "--config" ) ) || ( argc < 3 ) ) {
            pSvr = "<host:port> <Svc> <pubTmr> [<bDisk>]";
            Console.WriteLine( pSvr );
            return 0;
         }

         // [ <hosts> <Svc> <pubTmr> [<bDisk>]

         pSvr = args[0];
         pPub = args[1];
         tTmr = Convert.ToDouble( args[2] );
         bDsk = ( argc > 3 );
         app  = new SysAgent( pSvr, pPub, tTmr, bDsk );
         Console.WriteLine( "Hit <ENTER> to terminate..." );
         Console.ReadLine();
      } catch( Exception e ) {
         Console.WriteLine( "Exception: " + e.Message );
      }
      return 0;
   }

}  // class SysAgent
